### ISSUE TITLE
Start conditions + Nomenclature for higher number of tasks - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1840]
In GitLab by @dbeltrankyl on Jul 12, 2023, 08:58

_Merges DestinE-Start-Conditions -> master_

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 12, 2023, 09:00

Commented on [autosubmit/job/job.py line 1525](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..ccf4695375470cd50ceb1ffe4725274371997136#diff-f8751fbf5d2712919c09b25f56122c3fR1525)



To change when the merge is completed

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 12, 2023, 09:04

added 1 commit

<ul><li>db916120 - comment change</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15402&start_sha=280e4e78ba1b2937cb5a31dacf48237f27708c4d)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 13, 2023, 08:54

added 1 commit

<ul><li>83a26a41 - apply_parent changed to use the same function than in the other place</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15422&start_sha=db9161202401301212034ee05cc525bf902fceb6)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 13, 2023, 09:01

added 1 commit

<ul><li>ccf46953 - added additional tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15424&start_sha=83a26a4102cd2fcb48b3be67ca92b7648dc95ad2)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 13, 2023, 12:50

Commented on [autosubmit/job/job.py line 1525](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..ccf4695375470cd50ceb1ffe4725274371997136#diff-f8751fbf5d2712919c09b25f56122c3fR1525)



changed this line in [version 5 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15427&start_sha=ccf4695375470cd50ceb1ffe4725274371997136#c77b5e15aaf69b5ff74b6df825737a6687ced32b_1525_1528)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 19, 2023, 15:57

added 1 commit

<ul><li>9da19f52 - Added failure tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15538&start_sha=055c43bcbedf10c3abb6dcfc02c05a83772acb25)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 20, 2023, 08:54

marked this merge request as **ready**

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 20, 2023, 08:59

Hello @kinow 

Some docs are missing ( split section image)  but  merge should be ready to review, 

Testing suite unit passing, 

This contains:

* Start conditions ( STATUS ) 
* Triggers ( FROM STEP )
* [start:end:step] nomenclature
* [start:end:step] nomenclature also for X_to
* combine different nomenclatures ej [1:10:2]*,11,12*,[13:20] 
* "?" to mark the selection of a job
* "*" to mark 1:1 dependency 
* Colors to dependencies in the graph
* Numbers with the from_step value in the dependency graph

I'll try to expand a copy of the Francesc workflow later to do a real test as there are many variables to look at and maybe I'm missing something more

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 20, 2023, 08:59

requested review from @kinow

**Comment by kinow** [comment]
Review 90% done! Just going through the last two files :thumbsup:

**Comment by kinow** [comment]
I think we can use formatting for Sphinx here, so double back-ticks, like `\`\`job.type\`\``.

Also, the docs of all other variables end with a period, could you add one here too, please?

**Comment by kinow** [comment]
Should it be `sleep 5` as in the others (R/Python), or is there a reason for a longer sleep time here (maybe it's easier to see that checkpoints work that way?)

**Comment by kinow** [comment]
Not sure if s/upton/up to, upon, or something else?

**Comment by kinow** [comment]
Great job on the docs! Thank you a lot!!

**Comment by kinow** [comment]
Maybe worth adding a note whether they are case sensitive/insensitive.

**Comment by kinow** [comment]
Good to know!

**Comment by kinow** [comment]
I think this line should stay at the top near the rest of the license?

**Comment by kinow** [comment]
I know this is not the focus on this MR, but thank you a lot for slowly organizing the imports and the formatting in the code :bow: Much appreciated!

**Comment by kinow** [comment]
Partial review, @dbeltrankyl , as I think you will start working on another improvement here.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 20, 2023, 10:57

Commented on [autosubmit/job/job.py line 1528](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..37b15ceae9ade3cb1226687345fd8ce5ae06c221#diff-f8751fbf5d2712919c09b25f56122c3fR1528)



Oh, yes 

It is was there for testing proposes , will change it back

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 24, 2023, 16:05

Commented on [autosubmit/job/job.py line 1528](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..37b15ceae9ade3cb1226687345fd8ce5ae06c221#diff-f8751fbf5d2712919c09b25f56122c3fR1528)



changed this line in [version 11 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15657&start_sha=37b15ceae9ade3cb1226687345fd8ce5ae06c221#c77b5e15aaf69b5ff74b6df825737a6687ced32b_1528_1528)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 24, 2023, 16:05

added 1 commit

<ul><li>38420aab - fix offset</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15657&start_sha=37b15ceae9ade3cb1226687345fd8ce5ae06c221)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:01

Commented on [autosubmit/job/job_common.py line 134](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..38420aab2ccb5570eedec826f76caf3b76b58499#diff-dd9d9a2bd7b3349ad896e1c11dbb1a52R134)



changed this line in [version 12 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15678&start_sha=38420aab2ccb5570eedec826f76caf3b76b58499#384730a3f08e748140d00b594c7de5d0c02eff67_134_134)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:01

Commented on [docs/source/userguide/defining_workflows/index.rst line 215](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..38420aab2ccb5570eedec826f76caf3b76b58499#diff-90e604412c7ac0fa78cc8175c13ff674R215)



changed this line in [version 12 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15678&start_sha=38420aab2ccb5570eedec826f76caf3b76b58499#bb55553b9d40085a7ba16b894d2065afc93139e4_215_215)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:01

Commented on [autosubmit/job/job_list.py line 23](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..38420aab2ccb5570eedec826f76caf3b76b58499#diff-bad488e977334d250f2df049877c45f3R23)



changed this line in [version 12 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15678&start_sha=38420aab2ccb5570eedec826f76caf3b76b58499#e45075a97470bd3f004d8825e5560953e3c45ee8_23_23)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:01

added 1 commit

<ul><li>74e0a2a6 - Fixed few things , fixed docs, fixed and modify some tests</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15678&start_sha=38420aab2ccb5570eedec826f76caf3b76b58499)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:02

Okay, all tested with real examples seem to work fine( after more tweaks ). Test passing, docs passing.., and comments solved ( thanks for that). I think it is finally ready to merge without issues.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 25, 2023, 16:02

Commented on [autosubmit/job/job_list.py line 23](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..38420aab2ccb5570eedec826f76caf3b76b58499#diff-bad488e977334d250f2df049877c45f3R23)



pycharm refactor did this for no reason, didn't notice thanks

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 09:54

added 3 commits

<ul><li>1255b823 - todo</li><li>37a9d1b7 - 1-to-N and n-to-1 implemented</li><li>2cd9f34a - Merge branch &#39;improve_nomenclature_1_to_1&#39; into &#39;DestinE-Start-Conditions&#39;</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15808&start_sha=74e0a2a6799555e26daeb238dd8eecd9fc1c4968)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 09:54

merged 1-to-N and n-to-1 into this branch, docs added

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 09:55

I think this is ready to be merged in master @kinow

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 10:20

added 1 commit

<ul><li>70f05463 - Added tests and missing doc</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15812&start_sha=2cd9f34a1e011cc521927aca28285554f93d88ea)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 10:22

Doc and (1-to-N , N-to-1) were only in local ( I did make HTML, and In my head, it meant that it was pushed, sorry ) now it is really ready to merge

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 10:23

added 1 commit

<ul><li>53cf7330 - n-to-1 change test to precise syntax</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15814&start_sha=70f05463c0b13f1c6bae5a00c439c7209f5d5000)

**Comment by kinow** [comment]
Maybe this could be 

```
remote_checkpoint_file = f"{self.get_files_path()}/CHECKPOINT_{str(job.current_checkpoint_step)}"
```

Then use that in the rest of the code below (looks like `remote_checkpoint_path+str(job.current_checkpoint_step)` is duplicated in a few places).

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/190769b569d61356eefd61e4c5ad7737/image.png)

**Comment by kinow** [comment]
40% done. Going to check out the code and look at the rest with the IDE, and also try to test it with a local workflow. But it's looking good so far. Thanks @dbeltrankyl !

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 13:24

mentioned in issue ces/automatic_performance_profiling#60

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 13:47

Commented on [autosubmit/platforms/platform.py line 526](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..9e843d8212a970dbd960afeef0d0d883880af563#diff-5d19b70fc06bfdd8c8b4686d4aa7fc24R526)



Sure, I think originally the function was different and seems like a leftover thanks for reporting it

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 31, 2023, 14:33

Commented on [autosubmit/platforms/platform.py line 526](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..9e843d8212a970dbd960afeef0d0d883880af563#diff-5d19b70fc06bfdd8c8b4686d4aa7fc24R526)



Hello again, 

I remember why I did that in this version when I performed the change and not the reverse.

 In each iteration, we change the job.current_check_step value, so with your change, `remote_checkpoint_file`  would be fixed to the first one, no?

**Comment by kinow** [comment]
Reviewing it again now :tada: sorry the delay.

**Comment by kinow** [comment]
I think we won't need the setter, Dani. We can have just the `@property.getter` for `checkpoint`, and derive the value based on `self.type`.

Something like this:

```diff
diff --git a/autosubmit/job/job.py b/autosubmit/job/job.py
index 7bf33b5d..c12eb6cd 100644
--- a/autosubmit/job/job.py
+++ b/autosubmit/job/job.py
@@ -186,7 +186,6 @@ class Job(object):
         self.x11 = False
         self._local_logs = ('', '')
         self._remote_logs = ('', '')
-        self._checkpoint = None
         self.script_name = self.name + ".cmd"
         self.status = status
         self.prev_status = status
@@ -254,18 +253,13 @@ class Job(object):
     @property
     @autosubmit_parameter(name='checkpoint')
     def checkpoint(self):
-        """Generates a checkpoint step for this job based on job.type"""
-        return self._checkpoint
-
-    @checkpoint.setter
-    def checkpoint(self):
-        """Generates a checkpoint step for this job based on job.type"""
+        """Generates a checkpoint step for this job based on ``job.type``."""
         if self.type == Type.PYTHON:
-            self._checkpoint = "checkpoint()"
+            return "checkpoint()"
         elif self.type == Type.R:
-            self._checkpoint = "checkpoint()"
-        else: # bash
-            self._checkpoint = "as_checkpoint"
+            return "checkpoint()"
+        else:  # bash
+            return "as_checkpoint"
 
     def get_checkpoint_files(self):
         """
```

Also added a period as other variables description end with `.`, and tested the double backticks:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/38c4bde89faf16dd517b27302f950f5d/image.png)

**Comment by kinow** [comment]
Not sure if this is needed, `as_conf.jobs_data.get(self.section,{})`.

A few lines above we call `parameters['DEPENDENCIES'] = str(as_conf.jobs_data[self.section].get("DEPENDENCIES",""))`, I think that is already using `jobs_data[self.section]`... so it should be safe to use it directly here... I think.

**Comment by kinow** [comment]
:thumbsup: tested creating /tmp/test.sh

```bash
#!/bin/bash

AS_CHECKPOINT_CALLS=0

job_name_ptrn="JOB_1"

function as_checkpoint {
  AS_CHECKPOINT_CALLS=$((AS_CHECKPOINT_CALLS+1))
  touch ${job_name_ptrn}_CHECKPOINT_${AS_CHECKPOINT_CALLS}
}

echo ${AS_CHECKPOINT_CALLS}

as_checkpoint

echo ${AS_CHECKPOINT_CALLS}
```

It executed fine, and `shellcheck` only complained when I `shellcheck -o all test.sh`, but should be fine :thumbsup:

**Comment by kinow** [comment]
Do we need a `global` here, @dbeltrankyl ?

Tested by creating `/tmp/test.py`, and got an error below.

```python
AS_CHECKPOINT_CALLS = 0
def as_checkpoint():
  AS_CHECKPOINT_CALLS = AS_CHECKPOINT_CALLS + 1
  print(AS_CHECKPOINT_CALLS)


as_checkpoint()

as_checkpoint()

```

```bash
Traceback (most recent call last):
  File "/tmp/checkpoint/test.py", line 7, in <module>
    as_checkpoint()
  File "/tmp/checkpoint/test.py", line 3, in as_checkpoint
    AS_CHECKPOINT_CALLS = AS_CHECKPOINT_CALLS + 1
UnboundLocalError: local variable 'AS_CHECKPOINT_CALLS' referenced before assignment
```

**Comment by kinow** [comment]
Here variable hoisting works fine, I think only the Python example may need the `global` thingy.

```bash
kinow@ranma:/tmp/checkpoint$ cat test.R
AS_CHECKPOINT_CALLS = 0

as_checkpoint <- function() {
  AS_CHECKPOINT_CALLS <<- AS_CHECKPOINT_CALLS + 1
  print(AS_CHECKPOINT_CALLS)
}

as_checkpoint()

as_checkpoint()

kinow@ranma:/tmp/checkpoint$ Rscript test.R
[1] 1
[1] 2

```

**Comment by kinow** [comment]
s/and found_member is False/and not found_member ?

**Comment by kinow** [comment]
Maybe we can even remove this line.

**Comment by kinow** [comment]
Missing spaces here ? `,child=None,parent=None` (you fixed it in other parts on this same line :laughing: )

I spent some time with the IDE looking at this function and the warnings in pycharm. I don't know if these values are possible to exist in Autosubmit (they may fail to pass validation in some other part of the code, for example). But here are some examples that produce errors, Dani:

```python
from autosubmit.job.job_list import JobList

JobList._apply_filter(
    parent_value='bla',
    filter_value="bla*bla\\2",
    associative_list=None,
    level_to_check='DATES_FROM')

```

produces

```bash
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/autosubmit/test.py", line 3, in <module>
    JobList._apply_filter(
  File "/home/kinow/Development/python/workspace/autosubmit/autosubmit/job/job_list.py", line 450, in _apply_filter
    if child.split == parent_value:
AttributeError: 'NoneType' object has no attribute 'split'
```

And

```python
from autosubmit.job.job_list import JobList
from unittest import TestCase
from mock import Mock, patch

child = Mock()
child.split = 42


class TestBasicConfig(TestCase):

    @patch('autosubmit.job.job_list.JobList')
    def test_something(self, mocked: Mock):
        mocked._parse_filters_to_check.return_value = ['NATURAL']
        r = JobList._apply_filter(
            parent_value='bla',
            filter_value="bla*bla\\2",
            associative_list=[],
            level_to_check='DATES_FROM',
            child=child)
        print(f'The returned boolean is {r}')

```

Causes the function to return `None` instead of a boolean (not sure if that breaks anything or makes any difference).

**Comment by kinow** [comment]
PyCharm is warning for `lesser_value` possibly being called before being initialized here.

In an `else` block above you set `to_look_at_lesser = associative_list` and `lesser_group = -1`. Do we need also a value for `lesser_value` there? I think `greater` too. Maybe `lesser_value = None` and `greater = -1`?

**Comment by kinow** [comment]
Can the date be in another format? Like year only, `2004`? Probably not, but asking just in case...

**Comment by kinow** [comment]
>In each iteration, we change the job.current_check_step value, so with your change, `remote_checkpoint_file` would be fixed to the first one, no?

Each iteration in the `while`? I cannot see `remote_checkpoint_path` being changed there, Dani. Only after the `if job.current_checkpoint_step < job.max_checkpoint_step`?

**Comment by kinow** [comment]
``%AS_CHECKPOINT%`` in rst returns something like `%AS_CHECKPOINT%` in Markdown, @dbeltrankyl .

The values with single backtick here are displayed as italic when rendered :point_up: 

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/596d5171b871e97e30d24f6a837e24e2/image.png)

Changing that to double backticks,

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/7eb120b857cd06967b83dc39814b3c53/image.png)

**Comment by kinow** [comment]
@dbeltrankyl quick help learning how to use the `STATUS` + `FROM_STEP`, please :pray: 

I am trying to implement a pattern that's common with ecFlow & Cylc.

I built the basic diamond pattern with A{1} -> B{1,2} -> C{1}.

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES: B
```

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/187a25ba1a34ba0da133590112dfa947/image.png)

Now, what I want to do is to create a task that runs only if `A003_2_B` failed.

I tried with this, but it didn't produce what I wanted.

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES: B
  RECOVER_B_2:
    FILE: fix_b
    RUNNING: once
    DEPENDENCIES:
      B:
        STATUS: "FAILED"
        FROM_STEP: 1
```

it produced

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/6949e42526c23d05371dfca8a64b71c6/image.png)

There are two dotted lines. I take it these are start conditions? The color yellow seem to indicate the COMPLETED status, I think? Even though I requested `FAILED`?

Any idea what I'm doing wrong, please?

Tried `FAILED` without quotes, and also `FROM_STEP: 2`, same result.

Tagging @mcastril and @franra9 in case they'd like to follow the discussion here.

**Comment by kinow** [comment]
Another question @dbeltrankyl , what about invalid states? I tried changing the value and it didn't complain and produced the same graph above for this

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES: B
  RECOVER_B_2:
    FILE: fix_b
    RUNNING: once
    DEPENDENCIES:
      B:
        STATUS: "parangaricutirimicuaro"
        FROM_STEP: 1
```

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/9c42a3bb6e2de945acef9a25785f4f08/image.png)

What is stranger is that if I make a mistake of writing something different than `STATUS`, then I get a different arrow in the graph (it won't complain about the invalid value as we haven't implemented schema-salad/json-schema/any validation yet).

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES: B
  RECOVER_B_2:
    FILE: fix_b
    RUNNING: once
    DEPENDENCIES:
      B:
        # error here
        STATUS_PLAN: "parangaricutirimicuaro"
        FROM_STEP: 1
```

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/8ea8f13a1cf59b7974b96bf1e0e0a204/image.png)

**Comment by kinow** [comment]
Can we expand a little more the example here? Maybe we can replace "the task will be run if the dependency is running or any status after it." by being explicit which statuses will be ignored.

Or just add another sentence after the period saying which statuses are ignored. Reading it, I am not sure whether you are talking about the statuses above `RUNNING` or below :cold_sweat:

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:24

Commented on [autosubmit/platforms/platform.py line 526](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..9e843d8212a970dbd960afeef0d0d883880af563#diff-5d19b70fc06bfdd8c8b4686d4aa7fc24R526)



Hello, 

The remote_checkpoint_path is not updated but

In each iteration, the **job.current_checkpoint_step** is being increased.

Your request was to change remote_checkpoint to `remote_checkpoint_file = f"{self.get_files_path()}/CHECKPOINT_{str(job.current_checkpoint_step)}"`

But in that case, the while will always look for the first value, no? I want to do this

```
enter: checkpoint_0
while: checkpoint_0
remove(checkpoint_0)
get(checkpoint_1)
while(checkpoint_1)
...
```
Where the checkpoint_ is static, but the number isn't

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:39

Commented on [autosubmit/job/job.py line 1455](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..53cf73303190ef39d3984320f883d72857f4e3b0#diff-f8751fbf5d2712919c09b25f56122c3fR1455)



changed this line in [version 16 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15866&start_sha=53cf73303190ef39d3984320f883d72857f4e3b0#c77b5e15aaf69b5ff74b6df825737a6687ced32b_1455_1449)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:39

Commented on [autosubmit/job/job_common.py line 211](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..ced293ac87d181453c30ba0b733964cd66d738eb#diff-dd9d9a2bd7b3349ad896e1c11dbb1a52R211)



Fixed, It was missing the global for as_checkpoints and job_name_pttrn , thanks you!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:43

Commented on [autosubmit/job/job_list.py line 138](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..ccf4be61a5b6bd8f5eda39642729639c166741c1#diff-bad488e977334d250f2df049877c45f3R138)



I didn't touch this function. I think the refactor of Pycharm touched this part. but yes, it should be a not.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:43

Commented on [autosubmit/job/job_list.py line 138](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..ccf4be61a5b6bd8f5eda39642729639c166741c1#diff-bad488e977334d250f2df049877c45f3R138)



changed this line in [version 17 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15868&start_sha=ccf4be61a5b6bd8f5eda39642729639c166741c1#e45075a97470bd3f004d8825e5560953e3c45ee8_138_138)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:43

added 1 commit

<ul><li>1cb0171e - changed is false for not</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15868&start_sha=ccf4be61a5b6bd8f5eda39642729639c166741c1)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:44

Commented on [autosubmit/job/job_list.py line 363](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..ced293ac87d181453c30ba0b733964cd66d738eb#diff-bad488e977334d250f2df049877c45f3R363)



Yes, well I think in the other branch ( workflow optimization) is not there anymore but we can also change it there as the other branch is for 4.1

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:53

Commented on [autosubmit/job/job_list.py line 390](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R390)



```python
from autosubmit.job.job_list import JobList

JobList._apply_filter(
    parent_value='bla',
    filter_value="bla*bla\\2",
    associative_list=None,
    level_to_check='DATES_FROM')
```

I'm assuming that:

* this value is always an integer as 
* \\, which activates the 1-to-N, N-to-1 makes sense only for the splits ( as they can be different between sections) meanwhile the number of dates, members and chunks are equal for all the sections

I did not do much work related to yaml validation as there is another issue for that but I'll add code to just discard "\\" and do a normal 1-to-1

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 08:54

Commented on [autosubmit/job/job_list.py line 390](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R390)



> Causes the function to return `None` instead of a boolean (not sure if that breaks anything or makes any difference).

I think it is related, as the filter set is DATES_FROM

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:10

Commented on [autosubmit/job/job_list.py line 456](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R456)



Strangely, for me, the warning is not shown. I'll revise why the hint is not working in my pycharm

When there is no lesser_value, it doesn't matter if the lesser is parent or child as the relation will be 1-to-1 as there are the same number of splits..

I think I did not include a test to 1-to-N with equal splits. 

I'll put it in that case.

```python
lesser = str(parent_splits)
greater = str(child_splits)
lesser_value = "parent"            
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:29

Commented on [autosubmit/job/job_list.py line 390](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R390)



Well, I was wrong, I did a tunnel vision for splits and forgot to add: "if parent and child" to avoid get into these routines for the other filters when "*\\" is set

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:36

Commented on [autosubmit/job/job_list.py line 390](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R390)



changed this line in [version 18 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15870&start_sha=1cb0171eb762cb57502f0bfb2be1eb34f7767a74#e45075a97470bd3f004d8825e5560953e3c45ee8_390_389)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:36

Commented on [autosubmit/job/job_list.py line 456](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R456)



changed this line in [version 18 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15870&start_sha=1cb0171eb762cb57502f0bfb2be1eb34f7767a74#e45075a97470bd3f004d8825e5560953e3c45ee8_456_450)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:36

added 1 commit

<ul><li>b7e9cb6e - fixed two issues</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15870&start_sha=1cb0171eb762cb57502f0bfb2be1eb34f7767a74)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:37

Commented on [autosubmit/job/job_list.py line 617](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..ced293ac87d181453c30ba0b733964cd66d738eb#diff-bad488e977334d250f2df049877c45f3R617)



No, AS will crash earlier

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:38

Commented on [docs/source/userguide/defining_workflows/index.rst line 234](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..b7e9cb6e3d889cdcc84a7f93e14492de5a57dc8e#diff-90e604412c7ac0fa78cc8175c13ff674R234)



ah didn't realized it, thanks for telling me

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:39

Commented on [docs/source/userguide/defining_workflows/index.rst line 234](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..b7e9cb6e3d889cdcc84a7f93e14492de5a57dc8e#diff-90e604412c7ac0fa78cc8175c13ff674R234)



changed this line in [version 19 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15872&start_sha=b7e9cb6e3d889cdcc84a7f93e14492de5a57dc8e#bb55553b9d40085a7ba16b894d2065afc93139e4_234_234)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:39

added 1 commit

<ul><li>a1e23119 - added `%AS_CHECKPOINT%`</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15872&start_sha=b7e9cb6e3d889cdcc84a7f93e14492de5a57dc8e)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:40

Thanks, Bruno, for the review! I think I revised all the commented and pushed the changes

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 09:41

Additionally, I fix [1:2] when it is set in a chunks_level as the chunks in the associative list are an integer, not str so It was not finding the boundaries

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 3, 2023, 09:55

Aside from that, I think the workflow shouldn't have yellow edges if there's no other colour than black, if we understand that yellow is COMPLETED. What would be the difference between the yellow and the black arrows?

More in general... what would be the difference between black and yellow? We also had conditional dependencies (dashed arrow) in black.

Surely I am forgetting something.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 10:06

oops, 

```
            if job in child.edge_info.get("FAILED",{}):
                color = self._table.get(Status.FAILED,None)
                label = child.edge_info["FAILED"].get(job.name,0)[1]
            elif job.name in child.edge_info.get("RUNNING",{}):
                color =  self._table.get(Status.RUNNING,None)
                label = child.edge_info["RUNNING"].get(job.name,0)[1]
```

The failed check didn't have the job.name as the rest 

> Aside from that, I think the workflow shouldn't have yellow edges if there's no other colour than black, if we understand that yellow is COMPLETED. What would be the difference between the yellow and the black arrows?

The default was to completed status; it will just be None,None

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 10:12

added 1 commit

<ul><li>77c71673 - fixed color</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15874&start_sha=a1e23119280ddf1ef57b898dd216a190edadd150)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 11:34

It will just ignore the STATUS. The default was wrongly set to have the color of the completed instead of just don't put a label.

Should be fixed now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 11:41

Commented on [docs/source/userguide/defining_workflows/index.rst line 231](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..77c71673ede5e8470a5827377fcb0569e704da9f#diff-90e604412c7ac0fa78cc8175c13ff674R231)



changed this line in [version 21 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15886&start_sha=77c71673ede5e8470a5827377fcb0569e704da9f#bb55553b9d40085a7ba16b894d2065afc93139e4_231_231)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 11:41

added 1 commit

<ul><li>f9cc9afe - update doc</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15886&start_sha=77c71673ede5e8470a5827377fcb0569e704da9f)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 11:43

added 1 commit

<ul><li>33d3d028 - update doc</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15888&start_sha=f9cc9afe3dff058edaabfa9eb3b49ece932624bb)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 11:43

Commented on [docs/source/userguide/defining_workflows/index.rst line 231](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..77c71673ede5e8470a5827377fcb0569e704da9f#diff-90e604412c7ac0fa78cc8175c13ff674R231)



update docs and added your example corrected

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 3, 2023, 15:47

Just to mention that I revised all the comments

**Comment by kinow** [comment]
Starting another review now :tada: thanks for updating it Dani!

**Comment by kinow** [comment]
s/\`STATUS\`/\`\`STATUS\`\`

Single backquote is for italics, https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#inline-markup-and-special-characters-e-g-bold-italic-verbatim

**Comment by kinow** [comment]
:clap: thank you!!!

**Comment by kinow** [comment]
Do you think we could allow int's too? I tried

`SPLIT_TO: 2` and it failed with

```bash
 [CRITICAL] argument of type 'int' is not iterable [eCode=7070]
```

**Comment by kinow** [comment]
Maybe we could add an example with `FROM_STEP` too?

**Comment by kinow** [comment]
Thanks a lot for updating it @dbeltrankyl ! 

Here's my updated example script, to execute a task only when a specific task fails,

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES: B
  RECOVER_B_2:
    FILE: fix_b
    RUNNING: once
    DEPENDENCIES:
      B:
        STATUS: FAILED
        SPLITS_TO: "2"
```

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/983230cbbfad6ca5e98b42d21a6b93de/image.png)

I think I'm done with the code, now looking at the documentation and trying a few examples to see if I grokked how it works, and also to review the docs.

I changed the jobs to have A -> B, and then B_1 goes to C, and B_2 on FAILED goes to RECOVER_B_2. I am also trying to use checkpoints, to force C to start as soon as the first checkpoint is reached in B_1 (it has three).

```yaml
JOBS:
  A:
    FILE: a
    RUNNING: once
    SPLITS: 1
  B:
    FILE: b
    RUNNING: once
    SPLITS: 2
    DEPENDENCIES: A
  C:
    FILE: c
    RUNNING: once
    SPLITS: 1
    DEPENDENCIES:
      B:
        SPLITS_TO: "1"
        FROM_STEP: 2
  RECOVER_B_2:
    FILE: fix_b
    RUNNING: once
    DEPENDENCIES:
      B:
        STATUS: FAILED
        SPLITS_TO: "2"
```

The script for `B` is:

```bash
echo "I AM B"
sleep 30
%AS_CHECKPOINT%
echo "ONE CHECKPOINT DONE!"
%AS_CHECKPOINT%
sleep 300
echo "SECOND CHECKPOINT DONE!"
%AS_CHECKPOINT%
sleep 120
echo "B IS DONE!"
```

The structure of the workflow looks OK.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/f20d26319e478d94fde6b204f1280b7b/image.png)

@dbeltrankyl here's what happens when the workflow has finished:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/2423d0ead8272946141f4662e9a74b70/image.png)

Here it looks like `RECOVER_B_2` was completed. And it was actually executed

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/d1f22674b4a7edd2d059f16ff20a2c73/image.png)

Shouldn't it have been called only if `B_2` status was set to `FAILED`? I need to check, but IIRC in Cylc that task would have been SKIPPED or just never executed.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/c04249237a1981a3957c5c1bec81d90c/image.png)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:05

Commented on [docs/source/userguide/defining_workflows/index.rst line 321](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..ced293ac87d181453c30ba0b733964cd66d738eb#diff-90e604412c7ac0fa78cc8175c13ff674R321)



Not sure, they could also put SPLITS_TO: "2,3,4" or "[1:2]" perhaps is better to keep it as string, allowing it to be an int will complicate the code I think

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:07

> @dbeltrankyl here's what happens when the workflow has finished:

This is expected. Currently, the STATUS keyword only indicates from which status the job can start, but it can continue if all parents are completed. 

In the future, we can add a target status also and separate STATUS to STATUS_START and STATUS_TARGET. And maybe make the default for failures as you mentioned 

The former conditional dependencies were also working as it is now.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:39

Commented on [docs/source/userguide/defining_workflows/index.rst line 215](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..33d3d0281c2b1d828208738a9a7a1b74894338ab#diff-90e604412c7ac0fa78cc8175c13ff674R215)



changed this line in [version 23 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15924&start_sha=33d3d0281c2b1d828208738a9a7a1b74894338ab#bb55553b9d40085a7ba16b894d2065afc93139e4_215_215)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:39

added 1 commit

<ul><li>b550b539 - docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15924&start_sha=33d3d0281c2b1d828208738a9a7a1b74894338ab)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:56

added 57 commits

<ul><li>b550b539...09c3361d - 23 commits from branch <code>master</code></li><li>09c3361d...8e1dace4 - 24 earlier commits</li><li>7c0d0642 - n-to-1 change test to precise syntax</li><li>d6ddae45 - Fixed part of reviews,</li><li>62dfc040 - changed is false for not</li><li>d9741609 - fixed two issues</li><li>132673a5 - added `%AS_CHECKPOINT%`</li><li>6272bc53 - fixed color</li><li>789e7295 - update doc</li><li>87428527 - update doc</li><li>632aca09 - docs</li><li>9e843d82 - Merge branch &#39;DestinE-Start-Conditions&#39; of...</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15926&start_sha=b550b5393d822af20c14ea624bb76c31d26fbe99)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 4, 2023, 12:57

rebased to add the configparser test fix

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 08:21

Hello @kinow ,

Could we finish this issue // Do you have anything left to check? I want to merge this asap as I will be on holiday next Monday. Thanks.

**Comment by kinow** [comment]
Ah, sorry, you are right about not changing the remote_checkpoint_file. We should still be able to reduce the string concatenation in that function. I think this could work.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/282b789d0bc3aafbc4811f488b17ffd4/image.png)

**Comment by kinow** [comment]
Suspected so, thanks!

**Comment by kinow** [comment]
Thanks!

**Comment by kinow** [comment]
Looking good now, thanks Dani!

**Comment by kinow** [comment]
Fair point, marking as resolved :thumbsup:

**Comment by kinow** [comment]
Doing another review and testing a few last things. Just want to simulate what I think @franra9 will need, and then compare one last time with ecFlow&Cylc's status triggers. I should have finished it in the next hour/hours.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 10:21

Commented on [autosubmit/platforms/platform.py line 526](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..9e843d8212a970dbd960afeef0d0d883880af563#diff-5d19b70fc06bfdd8c8b4686d4aa7fc24R526)



changed this line in [version 25 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15959&start_sha=9e843d8212a970dbd960afeef0d0d883880af563#e00ded6b6df8370954fc259c5e67a43145a32028_526_526)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 10:21

added 1 commit

<ul><li>56f8115d - changed to f-string</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15959&start_sha=9e843d8212a970dbd960afeef0d0d883880af563)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 10:23

Commented on [autosubmit/platforms/platform.py line 526](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..9e843d8212a970dbd960afeef0d0d883880af563#diff-5d19b70fc06bfdd8c8b4686d4aa7fc24R526)



Hello Bruno, 

I changed it to f-string, thanks for that 

but with your changes,  current_checkpoint is not updated,  so the while will always check for the first checkpoint and I want to check the next one in the next iteration,no?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 10:27

Commented on [autosubmit/job/job_list.py line 456](https://github.com/BSC-ES/autosubmit/compare/5a886b4d277c3fb0e60d82a3d0558d9d4544e252..1cb0171eb762cb57502f0bfb2be1eb34f7767a74#diff-bad488e977334d250f2df049877c45f3R456)



This was added, the other day forgot to mention that it was uploaded

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 10:28

thanks you!

**Comment by kinow** [comment]
I think somewhere around here we should mention what you told me on Slack, @dbeltrankyl 

>the splits_from is mandatory for the *

**Comment by kinow** [comment]
Oh, you are correct again!

**Comment by kinow** [comment]
approved this merge request

**Comment by kinow** [comment]
Nothing else to add apart from this comment. I will leave the YAML for my experiment and the screenshot of the plot. @dbeltrankyl thanks a lot for the patience here and in Slack! I have an eFlows4HPC meeting in a few minutes, but I will try to review the `autosubmit stats` now/today as well, so that that can be included in the next release.

Thanks!

**Comment by kinow** [comment]
The jobs YAML I used for testing:

```yaml
JOBS:
  SIM:
    FILE: sim
    PLATFORM: local
    RUNNING: CHUNK
    DEPENDENCIES: "SIM-1"
  DN:
    FILE: dn
    PLATFORM: local
    RUNNING: CHUNK
    SPLITS: 2
    DEPENDENCIES:
      SIM
  OPA:
    FILE: opa
    PLATFORM: local
    RUNNING: chunk
    SPLITS: 2
    DEPENDENCIES:
      DN:
        SPLITS_FROM:
          "[1:2]":
            SPLITS_TO: "[1:2]*"
    FOR:
      NAME: ['2TMAX', '2TMIN']
  URBAN:
    FILE: urban
    PLATFORM: local
    RUNNING: chunk
    SPLITS: 2
    DEPENDENCIES:
      OPA_2TMAX:
        SPLITS_FROM:
          "[1:2]":
            SPLITS_TO: "[1:2]*"
      OPA_2TMIN:
        SPLITS_FROM:
          "[1:2]":
            SPLITS_TO: "[1:2]*"
  MHM:
    FILE: mhm
    PLATFORM: local
    RUNNING: chunk
    SPLITS: 2
    DEPENDENCIES:
      OPA_2TMIN:
        SPLITS_FROM:
          "[1:2]":
            SPLITS_TO: "[1:2]*"
```

And the plot using this branch

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/100898a0a9697f3dea326152e551c83a/image.png)

cc @franra9

**Comment by kinow** [comment]
mentioned in issue digital-twins/de_340/workflow#56

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 14:52

Commented on [docs/source/userguide/defining_workflows/index.rst line 441](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..ced293ac87d181453c30ba0b733964cd66d738eb#diff-90e604412c7ac0fa78cc8175c13ff674R441)



Thanks Bruno! I've added that sentence

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 14:53

Commented on [docs/source/userguide/defining_workflows/index.rst line 443](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..56f8115d5ca89b8acca0a9fe673b0da1680bbff5#diff-90e604412c7ac0fa78cc8175c13ff674R443)



changed this line in [version 26 of the diff](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15985&start_sha=56f8115d5ca89b8acca0a9fe673b0da1680bbff5#bb55553b9d40085a7ba16b894d2065afc93139e4_443_443)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 14:53

added 1 commit

<ul><li>ced293ac - docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/352/diffs?diff_id=15985&start_sha=56f8115d5ca89b8acca0a9fe673b0da1680bbff5)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 14:55

enabled an automatic merge when the pipeline for ced293ac87d181453c30ba0b733964cd66d738eb succeeds

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 7, 2023, 15:00

mentioned in commit 03dde90b6d009d40ad4fbcf24b739e6839447ad2
